client/components/calypso-i18n-provider/index.tsx

```diff
@@ -2,22 +2,26 @@
  * External dependencies
  */
 import * as React from 'react';
-import { setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
-import { I18nProvider } from '@automattic/react-i18n';
+import { createI18n, setLocaleData as setWpI18nLocaleData } from '@wordpress/i18n';
+import { I18nProvider } from '@wordpress/react-i18n';
 import i18n from 'i18n-calypso';
 import { LocaleProvider, i18nDefaultLocaleSlug } from '@automattic/i18n-utils';
 
 const CalypsoI18nProvider: React.FunctionComponent = ( { children } ) => {
-	const [ localeData, setLocaleData ] = React.useState( i18n.getLocale() );
 	const [ localeSlug, setLocaleSlug ] = React.useState( i18n.getLocaleSlug() );
 
+	// Create a new @wordpress/i18n instance when the locale changes,
+	// as `setLocaleData` doesn't replace the entire locale data, but rather merges it with the existing one,
+	// which may lead to residue translations from previous locales to remain causing displaying mixed up translations.
+	const wpI18n = React.useMemo( () => createI18n( i18n.getLocale() ), [ localeSlug ] );
+
 	React.useEffect( () => {
 		const onChange = () => {
 			// Create a new object to ensure the provider sees the data has changed
 			const nextLocaleData = Object.assign( {}, i18n.getLocale() );
```

Comment:
Here we do not need the `Object.assign` any more. `nextLocaleData` is no longer stored to state or passed as a prop to `I18nProvider` as it used to in the old version. So we don't need to ensure that it's a different object instance.
```js
const nextLocaleData = i18n.getLocale();
```
should be sufficient.